packages/react/src/hooks/useIsClient.spec.ts

```diff
@@ -1,12 +1,70 @@
 import { renderHook } from '@testing-library/react'
-import { useIsClient } from '.'
+import { useState, useSyncExternalStore } from 'react'
+import { noop } from '../utils'
+import { useIsClient, useIsomorphicLayoutEffect } from '.'
 
 describe('useIsClient', () => {
-  it('should return true when client side painting start', () => {
-    const {
-      result: { current: isClient },
-    } = renderHook(() => useIsClient())
+  it('should return true in client side rendering', () => {
+    const returnedFirst = renderHook(() => useIsClient())
+    expect(returnedFirst.result.current).toBe(true)
+    returnedFirst.unmount()
+    const returnedSecond = renderHook(() => useIsClient())
+    expect(returnedSecond.result.current).toBe(true)
+  })
+  it("'s comparison with legacy useIsClientOnly", () => {
+    // check CSR environment first
+    expect(typeof document !== 'undefined').toBe(true)
+
+    let renderCount = 0
+    let chanceIsClientToBeFalse = false
+
+    const emptySubscribe = () => noop
+    const getSnapshot = () => true
+    const getServerSnapshot = () => false
+    const useIsClient = () => {
+      const isClient = useSyncExternalStore(emptySubscribe, getSnapshot, getServerSnapshot)
+      renderCount++
+      if (!isClient) {
+        chanceIsClientToBeFalse = true
+      }
+      return isClient
+    }
+    const { unmount } = renderHook(() => useIsClient())
+    expect(renderCount).toBe(1)
+    expect(chanceIsClientToBeFalse).toBe(false)
+    unmount()
+    renderHook(() => useIsClient())
+    expect(renderCount).toBe(2)
+    expect(chanceIsClientToBeFalse).toBe(false)
+  })
+  it('improve legacy useIsClientOnly', () => {
+    // check CSR environment first
+    expect(typeof document !== 'undefined').toBe(true)
+
+    let renderCount = 0
+    let chanceIsClientToBeFalse = false
+
+    /**
+     * @deprecated This is legacy useIsClientOnly
+     */
+    const useIsClientLegacy = () => {
+      renderCount++
+      const [isClient, setIsClient] = useState(false)
+      if (!isClient) {
+        chanceIsClientToBeFalse = true
+      }
+      useIsomorphicLayoutEffect(() => {
+        setIsClient(true)
+      }, [])
 
-    expect(isClient).toBe(true)
+      return isClient
+    }
+    const { unmount } = renderHook(() => useIsClientLegacy())
+    expect(renderCount).toBe(2)
+    expect(chanceIsClientToBeFalse).toBe(true)
+    unmount()
+    renderHook(() => useIsClientLegacy())
+    expect(renderCount).toBe(4)
+    expect(chanceIsClientToBeFalse).toBe(true)
```

Comment:
@TkDodo If you okay, I want to request you to review this useIsClient hook.
When I test this in Next.js, I have no issue too. I also saw [this tweet](https://x.com/TkDodo/status/1741068994981826947?t=XmG17etMUL2m0JFim03vqw&s=19) and [this article](https://tkdodo.eu/blog/avoiding-hydration-mismatches-with-use-sync-external-store) but since this is the interface you proposed, I'd really like to hear your thoughts. This improvement could be big impact for even our all product of my company(@toss) and all users of Suspensive libraries

ps. @gwansikk and I are preparing @suspensive/react-query everyday to support @tanstack/react-query as one of community resources. (Recently [we start to support not only TanStack Query v4, also v5 at once](https://suspensive.org/docs/react-query/tanstack-query-compatibility) by using postinstall script like [vue-demi](https://github.com/vueuse/vue-demi) used by @tanstack/vue-query too.) I'm learning a lot from you and TanStack Query and I rely on TanStack Query v4, v5 for all my projects. Thank you!